node/modules/chainindex.go

```diff
@@ -0,0 +1,105 @@
+package modules
+
+import (
+	"context"
+	"path/filepath"
+
+	"go.uber.org/fx"
+	"golang.org/x/xerrors"
+
+	"github.com/filecoin-project/go-address"
+	"github.com/filecoin-project/go-state-types/abi"
+
+	"github.com/filecoin-project/lotus/chain/events"
+	"github.com/filecoin-project/lotus/chain/messagepool"
+	"github.com/filecoin-project/lotus/chain/stmgr"
+	"github.com/filecoin-project/lotus/chain/store"
+	"github.com/filecoin-project/lotus/chain/types"
+	"github.com/filecoin-project/lotus/chainindex"
+	"github.com/filecoin-project/lotus/node/config"
+	"github.com/filecoin-project/lotus/node/modules/helpers"
+	"github.com/filecoin-project/lotus/node/repo"
+)
+
+func ChainIndexer(cfg config.ChainIndexerConfig) func(lc fx.Lifecycle, mctx helpers.MetricsCtx, cs *store.ChainStore, r repo.LockedRepo) (chainindex.Indexer, error) {
+	return func(lc fx.Lifecycle, mctx helpers.MetricsCtx, cs *store.ChainStore, r repo.LockedRepo) (chainindex.Indexer, error) {
+		if cfg.DisableIndexer {
+			log.Infof("ChainIndexer is disabled")
+			return nil, nil
+		}
+
+		sqlitePath, err := r.SqlitePath()
+		if err != nil {
+			return nil, err
+		}
+
+		// TODO Implement config driven auto-backfilling
+		chainIndexer, err := chainindex.NewSqliteIndexer(filepath.Join(sqlitePath, chainindex.DefaultDbFilename),
+			cs, cfg.GCRetentionEpochs, cfg.ReconcileEmptyIndex, cfg.MaxReconcileTipsets)
+		if err != nil {
+			return nil, err
+		}
+
+		lc.Append(fx.Hook{
+			OnStop: func(_ context.Context) error {
+				return chainIndexer.Close()
+			},
+		})
+
+		return chainIndexer, nil
+	}
+}
+
+func InitChainIndexer(lc fx.Lifecycle, mctx helpers.MetricsCtx, indexer chainindex.Indexer,
+	evapi EventHelperAPI, mp *messagepool.MessagePool, sm *stmgr.StateManager) {
+	ctx := helpers.LifecycleCtx(mctx, lc)
+
+	lc.Append(fx.Hook{
+		OnStart: func(_ context.Context) error {
+			indexer.SetIdToRobustAddrFunc(func(ctx context.Context, emitter abi.ActorID, ts *types.TipSet) (address.Address, bool) {
```

Comment:
Thoughts about incorporating https://github.com/filecoin-project/lotus/issues/11594 ? it seems unfortunate that we couldn't land https://github.com/filecoin-project/lotus/pull/11723 but I haven't been able to square in my mind which way is ultimately better. Both approaches seem to have reason to suck. We should consider this carefully before landing.

Reply:
@rvagg I really would like to incorporate that on a follow up PR if at all we decide to land it. So far, none of the complaints I have seen from heavy users of this API have been around the addressing of the emitter and so I would like for a default do nothing approach till there is a strong user motivation for it as it is a non trivial change as is evident from https://github.com/filecoin-project/lotus/pull/11723.

Reply:
I'd really like to resolve #11594 properly here; either close it out as WONTFIX or implement it. Currently I'm leaning to a WONTFIX because I don't see it hurting us other than in db size (which isn't nothing). Let's keep this as part of the discussion. A migration is going to be really painful if we end up wanting to do it. May as well do it when we're creating from scratch if we have to do it.

Reply:
@rvagg I would like to close it as a `WONTFIX `.

Reply:
After a sync discussion last week, we agreed that the way to resolve this was to (1) always store the actor id as an int column and (2) store the DelegtatedAddress if it has one as a blob but otherwise leave it null.

Then the #11594 issue goes away and we have peace and harmony in the world.

Reply:
In the process this `SetIdToRobustAddrFunc` could be renamed, it could just be `LoadActorFunc` and we handle the `DelegatedAddress` bit locally in the indexer code.

Reply:
done in 7b304a4b4e3b3924d70fecd01555b4157a4196bd I see! nice, the I'll comment on the rename separately